### PR TITLE
[IMP] sale: add products from catalog

### PR DESCRIPTION
**Pull request by hote-odoo**
Before this commit to add products to a SO we needed to add one product at a time and then reclick on the 'Add product' and repeat. For large SO this become cumbersome fast.
After this commit and action 'Add from Catalog' will be added that will allow to add products on batch from the catalog, making more easy to manage sales orders.

See odoo/enterprise#37915

Task - 3062080

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale/views/sale_product_kanban_views.xml [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+                                    <img t-att-src="kanban_image('product.product', 'image_128', record.id.raw_value)"
+                                         class="o_image_64_contain"
+                                         alt="Product"/>
+                               </div>
+                                <div class="oe_kanban_details p-2 d-flex flex-column">
+                                    <div class="o_kanban_record_top flex-column w-100 justify-content-between">
+                                        <div class="o_kanban_record_title" t-attf-style="{{(record.sol_qty.value > 0) ? 'color: white' : ''}}">
+                                            <strong><field name="name"/></strong>
+                                        </div>
+                                        <div class="o_kanban_product_details" t-attf-style="{{(record.sol_qty.value > 0) ? 'color: white' : ''}}">
```
Comment (hote-odoo) [inline addons/sale/views/sale_product_kanban_views.xml:51]: The attribute `t-attf-style` is duplicated, I tried to set it as a value to avoid duplication but it does not work, it loses the reactivity.

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    setup() {
+        super.setup();
+        this.orm = useService("orm");
+
+        const refName = 'numpadDecimal';
+        useAutofocus({ refName });
+
+        this.state = useState({
+            readonly: this.props.readonly,
+        });
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:20]: I'm not sure for this part. Why do you need to change the readonly ?

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            this.props.update(this.props.value - 1);
+        }
+   }
+
+    addQuantity() {
+        this.props.update(this.props.value + 1);
+    }
+
+    // needed to trigger the onInput function of the useInputField in the FloatField Component
+    onInput(_) {}
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:42]: What do want to do ?

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        const refName = 'numpadDecimal';
+        useAutofocus({ refName });
+
+        this.state = useState({
+            readonly: this.props.readonly,
+        });
+    }
+
+    get formattedValue() {
+        return formatFloat(this.props.value, { noTrailingZeros: true });
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:24]: formatInteger

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
+/** @odoo-module */
+
+import { registry } from "@web/core/registry";
+import { useAutofocus, useService } from "@web/core/utils/hooks";
+import { formatFloat } from "@web/views/fields/formatters";
+import { FloatField } from '@web/views/fields/float/float_field';
+
+const { useState } = owl;
+
+export class SaleProductCatalogQuantity extends FloatField {
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:10]: IntegerField

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+/** @odoo-module */
+
+import { registry } from "@web/core/registry";
+import { IntegerField } from '@web/views/fields/integer/integer_field';
+
+
+export class SaleProductCatalogQuantity extends IntegerField {
+    setup() {
+        super.setup();
+    }
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:10]: remove

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    // update the quantity whenever the content is changed, otherwise the user
+    // have to click outside of the record to update the SOL quantity
+    onInput(ev) {
+        // otherwise we get a NaN bug when updating values from the keyboard
+        let value = parseInt(ev.target.value)
+        if (value >= 0) {
+            this.props.update(value);
+        }
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:29]: strange ?

Comment (hote-odoo) [reply addons/sale/static/src/js/sale_product_catalog_quantity.js:29]: when you clean the input through the keyboard it parses it as NaN and remove the input (as if the content is zero) but this is unwanted for an user that is removing the whole number to put a new one

File: addons/sale/static/src/js/sale_product_catalog_quantity.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    }
+}
+
+SaleProductCatalogQuantity.extractProps = ({ attrs }) => {
+    let integerProps = IntegerField.extractProps({attrs});
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog_quantity.js:30]: const

File: addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    // update the quantity whenever the content is changed, otherwise the user
+    // have to click outside of the record to update the SOL quantity
+    onInput(ev) {
+        // parse to avoid a NaN bug when updating values from the keyboard
+        let value = parseInt(ev.target.value)
+        if (value >= 0) {
+            this.props.update(value);
+        }
+    }
+}
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.js:27]: It s maybe easier to create your Field from scratch ?

File: addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.xml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        <xpath expr="input" position="after">
+            <button
+                name="sale_product_catalog_add_quantity"
+                class="m-1 btn d-flex align-items-center justify-content-center btn-secondary"
+                t-on-click.prevent="addQuantity">
+                <i class="fa fa-plus"/>
+            </button>
+        </xpath>
+    </t>
+
```
Comment (FrancoisGe) [inline addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.xml:27]: easier to create it from scratch

**Comment by hote-odoo** [comment]
Hello @tde-banana-odoo could you take a look to the `_track_finalize` method ? It seems to me to be no intrusive, but maybe I have to add something to it :)

File: addons/sale/models/mail_thread.py [hunk 11]
```diff
@@ -0,0 +1,17 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models
+
+class MailThread(models.AbstractModel):
+    _inherit = 'mail.thread'
+
+    def _track_finalize(self):
```
Comment (tde-banana-odoo) [inline addons/sale/models/mail_thread.py:9]: I would probably move this override on "sale.order" model, as normally method is called on the model. No need to override the global (abstract) method.

File: addons/sale/models/mail_thread.py [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+        # In the case sale orders are in draft mode do not log changes on the chatter.
+        # This is thought for sale orders while in form view or when in the catalog,
+        # in these cases only one sale order is selected.
+        if self._name == 'sale.order' and len(self) == 1 and self.state == 'draft':
```
Comment (tde-banana-odoo) [inline addons/sale/models/mail_thread.py:13]: To be honest, I feel like this is a hack for no real added value. Tracking values is standard in Odoo and I think it is better to stick on simple and reliable behavior, instead of adding magic. But that being said it is your choice to do it or not as it is localized on SO model.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing

File: addons/sale/models/mail_thread.py (hunk 12 above)
Comment (hote-odoo) [reply addons/sale/models/mail_thread.py:13]: @tde-banana-odoo It is desired for sale orders to not log changes on the chatter. From your help from discord this looks like the best way. If there is a more clean solution to it I will happily take it :)

**Review by chevalierv — CHANGES_REQUESTED** [review]
Hello @hote-odoo 

Thanks for your work on this :+1: This is a first pass :duck: 

Please ensure that none of my suggestions impact the design. I tried to get rid of useless classes. It could be a good idea to ask someone in design if they have any ideas to simplify or lighter the templates.

File: addons/sale/models/sale_order_line.py [hunk 13]
```diff
@@ -1183,3 +1183,39 @@ def _convert_to_sol_currency(self, amount, currency):
                 round=False,
             )
         return amount
+
+    def action_add_from_catalog(self):
+        order_id = self.env.context.get('order_id')
+        if not order_id:
+            raise UserError(_("Please save the sale order before adding from catalog."))
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/models/sale_order_line.py:1190]: Well it is apparently not needed anymore. When creating a new order and the order was not yet saved there was no `order_id` and the catalog crashed. I just tried and there is an `order_id` even when the order is not saved yet. I'm removing the `UserError`.

File: addons/sale/views/sale_product_kanban_views.xml [hunk 14]
```diff
@@ -0,0 +1,145 @@
+<odoo>
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/views/sale_product_kanban_views.xml:1]: I like it :)

File: addons/sale/views/sale_product_kanban_views.xml [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+                                                <field name="product_template_attribute_value_ids"
+                                                       widget="many2many_tags"
+                                                       groups="product.group_product_variant"
+                                                       options="{'color_field': 'color'}"/>
+                                            </div>
+                                        </div>
+                                    </div>
+                                </div>
+                                <div class="o_kanban_sale_bottom my-2 pt-3 border-top">
+                                    <div class="o_catalog_control text-dark">
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/views/sale_product_kanban_views.xml:57]: Text-dark is needed to see the numbers, otherwise the default to white and with the white background are difficult to see.

File: addons/sale/views/sale_product_kanban_views.xml [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+                                                <field class="me-2 text-white" name="priority" widget="priority"/>
+                                                <h4 class="o_kanban_record_title text-reset">
+                                                    <strong><field name="name"/></strong>
+                                                </h4>
+                                            </div>
+                                            <div class="o_kanban_product_details">
+                                                <div class="o_kanban_deault_code" t-if="record.default_code.value">
+                                                    [<field name="default_code"/>]
+                                                </div>
+                                                <div class="o_kanban_price">
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/views/sale_product_kanban_views.xml:42]: I need this one to find the div in the `sale_stock` module and add the quantities on hand info.

File: addons/sale_stock/views/sale_product_kanban_views.xml [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+<odoo>
+
+    <record id="sale_product_catalog_kanban_view_inherit" model="ir.ui.view">
+        <field name="name">sale.product.catalog.kanban.view.inherit</field>
+        <field name="model">product.product</field>
+        <field name="inherit_id" ref="sale.sale_product_catalog_kanban_view"/>
+        <field name="arch" type="xml">
+            <field name="default_code" position="after">
+                <field name="detailed_type" invisible="1"/>
+            </field>
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale_stock/views/sale_product_kanban_views.xml:11]: Same as before, you need to access the value in detailed type, the field need to be loaded before.

File: addons/sale/models/product_product.py [hunk 18]
```diff
@@ -42,6 +151,55 @@ def _onchange_type(self):
...
+                'sequence': ((sale_order.order_line and
+                                sale_order.order_line[-1].sequence + 1)
+                            or 10),  # put it at the end of the order
+            })
+            # computing name after creation,to benefit from sale_order_line logic
+            sol._compute_name()
+        elif qty != 0 and sale_order_line:
+            sale_order_line.write({'product_uom_qty': qty})
+        elif qty == 0 and sale_order_line:  # remove existing line if possible, otherwise set to zero
+            if sale_order_is_not_readonly:
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/models/product_product.py:186]: I reproduced the behavior from the sale order's form view state:
- when the so is in a state that does not allow removing lines (confirm, cancelled) the only thing we can do is set the quantity to zero;
- in the case we can remove the line then we do remove it.

File: addons/sale/models/product_product.py [hunk 19]
```diff
@@ -1,14 +1,123 @@
...
+        sale_order = self.get_sale_order()
+        if sale_order:
+            # only show the quantity if there is only one sol and the uom is the same
+            # in the sol and the product
+            for product_id, lines in groupby(
+                sale_order.order_line.filtered(lambda line: not line.display_type),
+                lambda line: line.product_id.id
+            ):
+                product = self.filtered(lambda product: product.id == product_id)
+                if len(lines) == 1 and product and lines[0].product_uom == product.uom_id:
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/models/product_product.py:37]: not yet resolved, I'm having and issue with the conversion, I will mark it resolved as fast as I can solve the bug

File: addons/sale/models/sale_order_line.py [hunk 20]
```diff
@@ -1163,3 +1163,29 @@ def _convert_to_sol_currency(self, amount, currency):
                 round=False,
             )
         return amount
+
+    def action_add_from_catalog(self):
+        order_id = self.env.context.get('order_id')
+        sale_order = self.env['sale.order'].browse(order_id).exists()
```
Comment (hote-odoo) [inline addons/sale/models/sale_order_line.py:1169]: I do this because if the sale order is not saved yet then `self.order_id` points to `False`.

Comment (hote-odoo) [reply addons/sale/models/sale_order_line.py:1169]: Forget this, I do not even need `sale_order`.

File: addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.js [hunk 21]
```diff
@@ -0,0 +1,48 @@
+/** @odoo-module */
+
+import { registry } from "@web/core/registry";
+import { integerField } from "@web/views/fields/integer/integer_field";
+
+export class SaleProductCatalogQuantity extends integerField.component {
+    async removeQuantity() {
```
[the remark this answers is not in the document]
Comment (hote-odoo) [reply addons/sale/static/src/js/sale_product_catalog/sale_product_catalog_quantity.js:8]: thank you !

File: addons/sale/models/product_product.py [hunk 22]
```diff
@@ -42,6 +153,54 @@ def _onchange_type(self):
...
+        order_id = self.env.context.get('order_id')
+        return self.env['sale.order'].search([('id', '=', order_id)])
+
+    def set_sol_qty(self, qty=1, click_on_record=True):
+        self.ensure_one()
+        # specify if clicked on the buttons of '+' and '-' or somewhere else on the record
+        self.sol_qty = self.sol_qty + 1 if click_on_record else qty
+
+    def remove_from_order(self):
+        self.sol_qty = 0
```
Comment (chevalierv) [inline addons/sale/models/product_product.py:202]: I'd like to handle that in js if possible, it's not really great to manage that on the server.

**Review by chevalierv — CHANGES_REQUESTED** [review]

**Comment by chapital** [comment]
Considering the catalog function,  can you show me please the alternatives products?, because right now only can use this function in website and POS modules, in case that this is not in, please keep in mind add the function, thanks

File: addons/sale/models/product_product.py (hunk 22 above)
Comment (chevalierv) [reply addons/sale/models/product_product.py:202]: See https://github.com/odoo-dev/odoo/pull/2469

**Comment by chapital** [comment]
@hote-odoo this is when we have a lot alternative products. Alternative products means that we have a products with same specs but different brand/internal reference(sku), this in references to not spend so many time searching the same product or even miss a sell, sometimes yes we have the product but in other brand. Just adding a button in the catalog view when a product have an alternative, also is important to see what products has stock in the catalog view.

**Comment by dalaOdoo** [comment]
@chapital thanks for the suggestion. 
Alternative products is currently a concept specific to website in Odoo so it will not be takein into account in this PR but the suggestion is good so I'll add it in feedback

File: addons/sale/controllers/catalog.py [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+                'readOnly': bool (optional)
+            }
+        """
+        order = request.env['sale.order'].browse(order_id)
+        sale_order_line_info = []
+        for product, lines in groupby(
+            order.order_line.filtered(lambda line: not line.display_type),
+            lambda line: line.product_id
+        ):
+            if (len(lines) > 1):
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:31]: Sooo, if i understand correctly:
If the product_id is not in the requested product_ids but there are multiple lines with the product, you'll load the information, but not if it's only present in one line ???

File: addons/sale/controllers/catalog.py [hunk 24]
```diff
@@ -0,0 +1,112 @@
...
+                    ),
+                    quantity=1.0,
+                    currency=lines[0].order_id.currency_id,
+                    date=lines[0].order_id.date_order,
+                )
+                sale_order_line_info.append(dict(
+                    productId=product.id,
+                    readOnly=True,
+                    price=price_unit,
+                ))
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:44]: You don't pop the product_id from product_ids in this case ?

File: addons/sale/controllers/catalog.py [hunk 25]
```diff
@@ -0,0 +1,112 @@
...
+                    lines[0].product_uom_qty, product.uom_id
+                )
+                price = order.pricelist_id._get_product_price(
+                    product=product.with_context(
+                        **lines[0]._get_product_price_context()
+                    ),
+                    quantity=quantity,
+                    currency=order.currency_id,
+                    date=order.date_order,
+                )
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:56]: Once again, why can't you take the price from the line (at least when the uoms are the same) ?

File: addons/sale/controllers/catalog.py [hunk 26]
```diff
@@ -0,0 +1,112 @@
...
+
+        :param int order_id: The sale order, as a `sale.order` id.
+        :param int product_id: The product, as a `product.product` id.
+        :param float quantity: The quantity selected in the product catalog.
+        :return: The unit price price of the product, based on the pricelist of the sale order and
+                 the quantity selected.
+        :rtype: float
+        """
+        sol = request.env['sale.order.line'].search([
+            ('order_id', '=', order_id), ('product_id', '=', product_id)
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:86]:
```suggestion
            ('order_id', '=', order_id), ('product_id', '=', product_id),
```

File: addons/sale/controllers/catalog.py [hunk 27]
```diff
@@ -0,0 +1,112 @@
...
+            order = request.env['sale.order'].browse(order_id)
+            sol = request.env['sale.order.line'].create({
+                'order_id': order.id,
+                'product_id': product_id,
+                'product_uom_qty': quantity,
+                'sequence': ((order.order_line and order.order_line[-1].sequence + 1) or 10),  # put it at the end of the order
+            })
+        elif quantity != 0 and sol:
+            sol.product_uom_qty = quantity
+        elif quantity == 0 and sol:
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:98]: Those if elif seem confusing to me, could maybe be reorganized (not 100% sure how though).

File: addons/sale/controllers/catalog.py [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+                ),
+                quantity=1.0,
+                currency=sol.order_id.currency_id,
+                date=sol.order_id.date_order,
+            )
+            if sol.order_id.state in ['draft', 'sent']:
+                sol.unlink()
+            else:
+                sol.product_uom_qty = 0
+            return price_unit
```
Comment (Feyensv) [inline addons/sale/controllers/catalog.py:111]: AFAIK, even if the qty is set to 0, the price_unit could be taken from the line (and therefore, you would only have to recompute the price with the pricelist call when the line is deleted).

File: addons/sale/models/sale_order.py [hunk 29]
```diff
@@ -1268,6 +1280,20 @@ def _create_invoices(self, grouped=False, final=False, date=None):
...
+        if (len(self) == 1
+            and self.env.cache.contains(self, self._fields['state'])
+            # The method _track_finalize is sometimes called too early or too late and it
+            # might cause a desynchronization with the cache, thus this condition is needed.
+            and self.state == 'draft'):
+            self.env.cr.precommit.data.pop(f'mail.tracking.{self._name}', {})
+            self.env.flush_all()
+            return
+        super()._track_finalize()
+
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:1296]: Should be in a separate commit with clear explanation

File: addons/sale/models/sale_order_line.py [hunk 30]
```diff
@@ -1163,3 +1163,28 @@ def _convert_to_sol_currency(self, amount, currency):
...
+        order_id = self.env.context.get('order_id')
+        kanban_view_id = self.env.ref('sale.sale_product_catalog_kanban_view').id
+        search_view_id = self.env.ref('sale.sale_product_catalog_search_view').id
+        domain = self.order_id._get_product_catalog_domain()
+
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Products'),
+            'res_model': 'product.product',
+            'views': [(kanban_view_id, 'kanban'), (False, 'form')],
```
[the remark this answers is not in the document]
Comment (Feyensv) [reply addons/sale/models/sale_order_line.py:1176]: why was this comment resolved without any change ?

File: addons/sale/models/sale_order_line.py [hunk 31]
```diff
@@ -1163,3 +1163,28 @@ def _convert_to_sol_currency(self, amount, currency):
...
+            'search_view_id': [search_view_id, 'search'],
+            'domain': order._get_product_catalog_domain(),
+            'context': {
+                **self.env.context,
+                'order_id': order.id,
+                'currency_id': order.currency_id.id,
```
Comment (Feyensv) [inline addons/sale/models/sale_order_line.py:1182]:
```suggestion
                'product_catalog_currency_id': order.currency_id.id,
```
to make sure there is no side effect from this context key

File: addons/sale/static/src/js/product_catalog/kanban_record.js [hunk 32]
```diff
@@ -0,0 +1,110 @@
+/** @odoo-module */
+import { useState, useSubEnv } from "@odoo/owl";
+import { useService } from "@web/core/utils/hooks";
+import { KanbanRecord } from "@web/views/kanban/kanban_record";
+import { useDebounced } from "@web/core/utils/timing";
+import { ProductCatalogSOL } from "./sale_order_line/sale_order_line"
+
+export class ProductCatalogKanbanRecord extends KanbanRecord {
+    static template = 'sale.ProductCatalogKanbanRecord';
```
Comment (Feyensv) [inline addons/sale/static/src/js/product_catalog/kanban_record.js:9]: Mix of single and double quoted strings in this file

File: addons/sale/static/src/js/product_catalog/sale_order_line/sale_order_line.xml [hunk 33]
```diff
@@ -0,0 +1,46 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<templates xml:space="preserve">
+    <t t-name="sale.ProductCatalogSOL" owl="1">
+        <div class="my-2 pt-3 border-top d-flex justify-content-end">
+            <t t-portal="'#product-'+props.productId">
```
Comment (Feyensv) [inline addons/sale/static/src/js/product_catalog/sale_order_line/sale_order_line.xml:5]: might be useful to add a comment here to explain the use of the t-portal (you explained it to me, but it should be clear by itself).

File: addons/delivery/models/product_product.py [hunk 34]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+from odoo import models
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    def _write_sale_order_line(self, sale_order, sale_order_line):
+        sale_order.write({'recompute_delivery_price': True})
+        super()._write_sale_order_line(sale_order, sale_order_line)
```
Comment (Feyensv) [inline addons/delivery/models/product_product.py:10]: Dead code

**Review by Feyensv — COMMENTED** [review]
Tech review, gonna play with the feature

File: addons/sale/models/sale_order_line.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/sale/models/sale_order_line.py:1176]: I thought this was useless but it is needed for the action edit in the kanban menu.

File: addons/sale/controllers/catalog.py [hunk 35]
```diff
@@ -0,0 +1,112 @@
...
+        ):
+            if (len(lines) > 1):
+                price_unit = lines[0].order_id.pricelist_id._get_product_price(
+                    product=lines[0].product_id.with_context(
+                        **lines[0]._get_product_price_context()
+                    ),
+                    quantity=1.0,
+                    currency=lines[0].order_id.currency_id,
+                    date=lines[0].order_id.date_order,
+                )
```
[the remark this answers is not in the document]
Comment (chevalierv) [reply addons/sale/controllers/catalog.py:39]: At this point, I need it because the uom can be different for each product, and also because I don't show the quantity so I assumed the price for 1 should be showed.

File: addons/delivery/models/product_product.py (hunk 34 above)
Comment (Feyensv) [reply addons/delivery/models/product_product.py:10]: Edit: should be handled by calling `onchange_order_line` in an override of the controller update route

File: addons/sale/models/sale_order_line.py [hunk 36]
```diff
@@ -1163,3 +1169,27 @@ def _convert_to_sol_currency(self, amount, currency):
                 round=False,
             )
         return amount
+
+    def action_add_from_catalog(self):
```
Comment (Feyensv) [inline addons/sale/models/sale_order_line.py:1173]: action_add_from_catalog is not really a `tooling`, should go in action methods.

File: addons/sale_stock/models/sale_order_line.py [hunk 37]
```diff
@@ -354,3 +354,8 @@ def _update_line_quantity(self, values):
         if line_products.mapped('qty_delivered') and float_compare(values['product_uom_qty'], max(line_products.mapped('qty_delivered')), precision_digits=precision) == -1:
             raise UserError(_('The ordered quantity cannot be decreased below the amount already delivered. Instead, create a return in your inventory.'))
         super(SaleOrderLine, self)._update_line_quantity(values)
+
+    def _get_action_add_from_catalog_extra_context(self, order):
+        extra_context = super()._get_action_add_from_catalog_extra_context(order)
+        extra_context.update(warehouse=order.warehouse_id.id)
+        return extra_context
```
Comment (Feyensv) [inline addons/sale_stock/models/sale_order_line.py:361]: Beautiful hook :1st_place_medal: 
(another way would have been to call `super(SaleOrderLine, self.with_context(warehouse=self.order_id.warehouse_id.id))....`)

**Comment by Feyensv** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
@robodoo rebase-ff
